### DISCUSSION TITLE
How do I get the response from a image protocol request?

### DISCUSSION THREAD

**Discussion by glyh** [kovidgoyal/kitty#7319] [Q&A]
For now it looks like everything is going into stdin from kitty's output, but when I do something like
```
  echo -n "${TERM_ESC}_Gi=123,f=100,t=f,a=T;$path_encoded$TERM_ESC\\"
read -r > response
```
This doesn't work.

**Comment by kovidgoyal** [comment]
you cannot do this in shell script. Write a proper program and read the
response from the terminal device.

**Reply by glyh** [reply]
What's preventing me from doing so?  @kovidgoyal

**Reply by glyh** [reply]
Also is there an example in python or something?

**Comment by koterpillar** [comment]
zsh can read output character by character, which is probably not that efficient but good enough. Here is an example:

```shell
read_kitty_response() {
  # Read the string like:
  # <ESC>_Gi=31;error message or OK<ESC>\
  escapes=0
  response=""
  while [ $escapes -lt 2 ]
  do
    read -r -s -k nextchar
    if [ "$nextchar" = $'\e' ]
    then
      escapes=$((escapes + 1))
    else
      response="$response$nextchar"
    fi
  done
  read -r -s -k nextchar  # final backslash
  echo "$response"
}

# Replace with the desired Kitty command
echo -n "${TERM_ESC}_Gi=123,f=100,t=f,a=T;$path_encoded$TERM_ESC\\"

echo "Kitty response is: $(read_kitty_response)"
```

This works for me in zsh (it has [-k for read](https://zsh.sourceforge.io/Doc/Release/Shell-Builtin-Commands.html)); for bash try `read -r -s -n 1 nextchar`.
